### DISCUSSION TITLE
How to ignore stationary false positives?

### DISCUSSION THREAD

**Discussion by raintonr** [blakeblackshear/frigate#15407] [Ask A Question] [answered]
I have previously asked about ignoring false positive that seem to suddenly appear in part of the frame and do not move (see #10414). The answer at the time was that logic was already improved in 0.14. I have now been running 0.14.1-f4f3cfa for some time and still see this problem.

False positives occur more often at night and/or in bad weather (when there can be rain or snow on the lens). While I get that false positives can occur, what I do not understand is that when they appear in frame and don't move, why are they causing detections, when I was told in the issue above that an object had to move to be considered a valid detection?

Here's a couple of examples:

![image](https://github.com/user-attachments/assets/681f9bb6-a2dc-41aa-9bd1-467bdb23bf87)

![image](https://github.com/user-attachments/assets/46638658-9962-48bb-8f24-de866fa6a2d3)

If Frigate does not flag objects as detections unless they move, it would be super useful to see the path that the object followed up to the point it was thought a valid target in order to investigate this further?

If the behaviour I am seeing is as designed, any suggestions as to how to avoid detections like this?

**Comment by raintonr** [comment]
BTW, I did think about increasing height and/or area necessary, but that would not work as sometimes the false positive is much larger:

![image](https://github.com/user-attachments/assets/14e6267d-21b9-4444-a865-12d9bd6ccb12)

**Comment by clsferguson** [comment]
you can increase the min_score and threshold, that may help with falses.

**Comment by NickM-27** [comment]
In 0.15 you'll be able to see the different moments an object was detected during its lifecycle to debug that

**Comment by raintonr** [comment]
Just wanted to thank you guys for the 'object lifecycle' stuff in 0.15 so can further investigate why I'm seeing false positives like this. Here's an example:

![image](https://github.com/user-attachments/assets/cad9c6fd-913a-42f3-9ae7-039e5be3bb34)

![image](https://github.com/user-attachments/assets/b6e10245-cc7e-4aed-8ec6-2cb4d8d50218)

![image](https://github.com/user-attachments/assets/aa38d350-f52f-4728-8efd-0b64bf027e4f)

Looking at this, the object's location (yellow dot bottom of the box) barely moves. I have previously been told that Frigate will ignore objects that don't move, but the definition of a 'moving object' seems too sensitive. Is there any way to tune this?

**Reply by NickM-27** [reply]
movement does not depend on the bottom center, it depends on the iou of the boxes. In this case the box changed considerably in each of the steps that you are showing, trying to filter this out based on movement is not the correct approach

**Reply by raintonr** [reply]
OK. What would you suggest then?

**Reply by NickM-27** [reply]
in this case a max_area filter likely makes sense as I don't see a cat being that large. Could also try changes to the camera itself to avoid these water drops from hitting it so much

**Reply by raintonr** [reply]
Yes, `max_area` would likely help with cats & dogs, but...

This is an intercom camera at a gate which can also imagine it's seeing people in the raindrops - `max_area` would not be able to help there, because when a person approaches the intercom they would likely cover a very large area.

This is why I asked in the previous ticket (#10414) about traversing the frame or passing through multiple zones.

What you are saying is that under the the rules Frigate uses to determine object motion, the example (https://github.com/blakeblackshear/frigate/discussions/15407#discussioncomment-12176325) qualifies. But surely you would also agree that if you showed these three frames to a person they would not consider the box to have moved or changed significantly at all?

Aren't there are some variables around these movement rules within Frigate & wouldn't a change to allow them to be adjusted by the user be possible?

**Reply by NickM-27** [reply]
> This is an intercom camera at a gate which can also imagine it's seeing people in the raindrops - max_area would not be able to help there, because when a person approaches the intercom they would likely cover a very large area.

Max area is object specific so that is fine, will help improve. 

> Aren't there are some variables around these movement rules within Frigate & wouldn't a change to allow them to be adjusted by the user be possible?

This is very internal to frigates tracking logic. Not only for initial object detection but also tracking, stationary / active detection, etc. users configuring these values would greatly change all of these behaviors and likely cause issues that users will have no idea how the changes impact it because you have to understand the code to understand the impacts. This is not something simple like motion detection threshold / contour area. 

And again, regardless, lack of motion does not say whether an object is a true or false positive. This is an object detection issue.

**Reply by NickM-27** [reply]
it's also worth pointing out that when a person walks directly away from or towards a camera their box changes very little, only getting slightly smaller / larger, this could cause a person to be ignored simply because the box changed little from frame to frame

**Reply by raintonr** [reply]
> Max area is object specific

Yes, I know. This is why I said it could help reduce false positives for objects that are expected to be small (like cats), but not for those that could almost fill the frame (like a person when they approach an intercom camera).

> ...users will have no idea how the changes impact it because you have to understand the code to understand the impacts...

This is a design/usability problem then. Good software hides complexity from the user and exposes useful functions in an intuitive and friendly manner.

> This is an object detection issue.

Indeed it is. I mean, look at the 'cat' in the example above. I simply cannot see where a 'cat' is visible in that box. Can you?

The main problem that I see with false positives when the camera lens is dirty/wet/etc. is that they seem to appear mid-frame, stay motionless (or at least very still as far as I can tell), then disappear. If trying to analyse motion is not the key, what is?

Does the model need improving? Do I need to submit all these false positives to Frigate+ and hope that a model without these issues can be trained?

**Reply by NickM-27** [reply]
> Good software hides complexity from the user and exposes useful functions in an intuitive and friendly manner.

I agree, which is exactly why the tracking specifics are not configurable, and was exactly my point for why it should remain that way. 

> stay motionless (or at least very still as far as I can tell), then disappear. If trying to analyse motion is not the key, what is?

You're hyper focused on this single instance and saying if you could make the object position changes logic less sensitive it would remove this false positive, but entirely ignoring the fact that this would also greatly affect other situations where the box of a true positive object does not change much from frame to frame leading to false negatives. 

I don't see a config posted as far as I can tell but you could try a larger model such as yolonas and see if these false positives go away, or yes subscribe to Frigate+ and work on training out these false positives if they occur.

**Reply by raintonr** [reply]
> You're hyper focused on this single instance...

Not really. I see false positives like this on multiple cameras mostly in rain/snow or when the lens needs cleaning. All I know is that a human would easily notice these are false and think one of the ways they would know would be the appearance mid-frame and lack of movement through the frame (yes - in my case - others might have cameras looking down a long corridor which I agree would need a different approach). If you say this doesn't translate into a method the Frigate detector could use then fair enough.

> I don't see a config posted...

I have a Coral Edge TPU in an old x86 box so looking at the docs seems YOLO-NAS isn't compatible.

How about looking at this another way...

During rain and snow it's inevitable cameras will get moisture on the lens and for whatever reason this is causing the detector to record false positives. If one could determine when these conditions are occurring then perhaps it would be useful to say, _"Hey, it's raining... turn up the `threshold` value for the detector as I know you can't see that well right now and want you to be sure before alerting me"_.

Hardly ideal, but that's life, right? When weather conditions are bad one cannot expect a security camera system to function as well as when visibility is perfect.

Unless I've missed something, Frigate detection thresholds cannot be dynamically adjusted, even for day/night (which is another example of good/bad visibility). I have seen the `improve_contrast` for day/night but this doesn't seem like the solution. Do you think adjustment of detection object thresholds, perhaps via MQTT topics, would be useful?

Or would you rather concentrate on just creating a better model (perhaps necessitating better hardware) to simply have more accurate detections in all conditions? I agree that would be ideal but how likely it is to happen?

**Reply by NickM-27** [reply, marked as answer]
We have some ideas for improving the motion detection to avoid detecting superfluous motion, but given that the moisture remains on the camera it is difficult to detect that.

Currently the thresholds can't be changed, I believe there is a pinned feature request for this to be possible to control.

> Or would you rather concentrate on just creating a better model (perhaps necessitating better hardware) to simply have more accurate detections in all conditions? I agree that would be ideal but how likely it is to happen?

all I can say is after training my Frigate+ model, I have literally not had a single false positive in many months. Many users have reported great results as well, but circumstances, camera environment, etc. certainly vary between users and that can lead to different results.

**Reply by raintonr** [reply]
OK, cool. That sounds positive then.

+1 for https://github.com/blakeblackshear/frigate/issues/2569
